File: Client-Server/server/predict.js

const {PythonShell} = require('python-shell')
const {parse_result} = require('./utils')

const predict = (data) => {
  const options = {
    args:[data]
  }
  PythonShell.run("predict.py",options,function(err,res) {
    if(err) {
      console.log('python-error: '+err);
    } else {
      // console.log('python-results '+res);
     console.log((parse_result(JSON.parse(res))))
    }
  })
}

const format_data = (data) => {
  // console.log(Object.entries(data));
  data = Object.fromEntries(Object.entries(data).map(([k,v]) => [k,{'0':v}]))
  // console.log(data);
  return data
}

// const data = JSON.parse('{"pkSeqID":{"0":3142762},"proto":{"0":"udp"},"saddr":{"0":"192.168.100.150"},"sport":{"0":"6551"},"daddr":{"0":"192.168.100.3"},"dport":{"0":"80"},"seq":{"0":251984},"stddev":{"0":1.900363},"N_IN_Conn_P_SrcIP":{"0":100},"min":{"0":0.0},"state_number":{"0":4},"mean":{"0":2.687519},"N_IN_Conn_P_DstIP":{"0":100},"drate":{"0":0.0},"srate":{"0":0.494549},"max":{"0":4.031619}}')

// const data = {
//     pkSeqID: "3142762",
//     proto: 'udp'
//   }
// console.log(format_data(data));
// const data_json = JSON.stringify(data)
// // console.log(data_json);
// predict(data_json)


console.log(new Date());
console.log(Date.now());
console.log(new Date().toString());
const obj = {
  client_id: 1,
  interval: [0,Date.now()]
};
console.log(JSON.stringify(obj));

// {
//   pkSeqID: { '0': 3142762 },
//   proto: { '0': 'udp' },
//   saddr: { '0': '192.168.100.150' },
//   sport: { '0': '6551' },
//   daddr: { '0': '192.168.100.3' },
//   dport: { '0': '80' },
//   seq: { '0': 251984 },
//   stddev: { '0': 1.900363 },
//   N_IN_Conn_P_SrcIP: { '0': 100 },
//   min: { '0': 0 },
//   state_number: { '0': 4 },
//   mean: { '0': 2.687519 },
//   N_IN_Conn_P_DstIP: { '0': 100 },
//   drate: { '0': 0 },
//   srate: { '0': 0.494549 },
//   max: { '0': 4.031619 }
// }

File: Client-Server/server/predict.py
import pickle
import pandas as pd
import numpy as np
import sys
from xgboost import XGBClassifier

class GradientBoost:
    def __init__(self):
        self.xgb_attack = XGBClassifier()
        self.xgb_category = XGBClassifier()
        self.xgb_subcategory = XGBClassifier()
    
    def fit(self,X_train,y_train):
        self.xgb_attack.fit(X_train,y_train['attack'])
        
        features_category = np.concatenate((X_train,np.array(y_train['attack']).reshape(-1,1)),axis=1)
        self.xgb_category.fit(features_category,y_train['category'])
        
        features_subcategory = np.concatenate((features_category,np.array(y_train['category']).reshape(-1,1)),axis=1)
        self.xgb_subcategory.fit(features_subcategory,y_train['subcategory'])
        
    def predict(self,X_test):
        predict_attack = self.xgb_attack.predict(X_test)
        
        test_category = np.concatenate((X_test,predict_attack.reshape(-1,1)),axis=1)
        predict_category = self.xgb_category.predict(test_category)
        
        test_subcategory = np.concatenate((test_category,predict_category.reshape(-1,1)),axis=1)
        predict_subcategory = self.xgb_subcategory.predict(test_subcategory)
        
        return pd.DataFrame({'attack':predict_attack,'category':predict_category,'subcategory':predict_subcategory})

filename = 'xgb_1.sav'
model = pickle.load(open(filename, 'rb'))

# data_json = '{"pkSeqID":{"0":3142762},"proto":{"0":"udp"},"saddr":{"0":"192.168.100.150"},"sport":{"0":"6551"},"daddr":{"0":"192.168.100.3"},"dport":{"0":"80"},"seq":{"0":251984},"stddev":{"0":1.900363},"N_IN_Conn_P_SrcIP":{"0":100},"min":{"0":0.0},"state_number":{"0":4},"mean":{"0":2.687519},"N_IN_Conn_P_DstIP":{"0":100},"drate":{"0":0.0},"srate":{"0":0.494549},"max":{"0":4.031619}}'
data_json = sys.argv[1]
data = pd.read_json(data_json)
ten_best_features = data[['seq','stddev','N_IN_Conn_P_SrcIP', 'min', 'state_number', 'mean', 'N_IN_Conn_P_DstIP','drate', 'srate', 'max']]
data_ten_best = data[ten_best_features.columns]
result = model.predict(data_ten_best)
print(result.to_json())
